bazel/generate_objc.bzl

```diff
@@ -0,0 +1,221 @@
+load(
+    "//bazel:protobuf.bzl",
+    "get_include_protoc_args",
+    "get_plugin_args",
+    "proto_path_to_generated_filename",
+)
+load(
+    ":grpc_util.bzl", 
+    "to_upper_camel_with_extension",
+    "label_to_file",
+)
+
+_GRPC_PROTO_HEADER_FMT = "{}.pbrpc.h"
+_GRPC_PROTO_SRC_FMT = "{}.pbrpc.m"
+_PROTO_HEADER_FMT = "{}.pbobjc.h"
+_PROTO_SRC_FMT = "{}.pbobjc.m"
+_GENERATED_PROTOS_DIR = "_generated_protos"
+
+_GENERATE_HDRS = 1
+_GENERATE_SRCS = 2
+_GENERATE_NON_ARC_SRCS = 3
+
+def _generate_objc_impl(ctx):
+
+    """Implementation of the generate_cc rule."""
+    protos = [
+        f
+        for src in ctx.attr.deps
+        for f in src[ProtoInfo].transitive_imports.to_list()
+    ]
+    outs = []
+
+    label_package = _join_directories([ctx.label.workspace_root, ctx.label.package])
+
+    files_with_rpc = [label_to_file(f) for f in ctx.attr.srcs]
+    for proto in protos:
+        outs += [_get_file_out_from_proto(proto, _PROTO_HEADER_FMT)]
+        outs += [_get_file_out_from_proto(proto, _PROTO_SRC_FMT)]
+
+        file_path = _strip_package_from_path(label_package, proto)
+        if not file_path.startswith("//"):
+            pass # TODO: check package boundary
+        if file_path in files_with_rpc or proto.path in files_with_rpc:
+            outs += [_get_file_out_from_proto(proto, _GRPC_PROTO_HEADER_FMT)]
+            outs += [_get_file_out_from_proto(proto, _GRPC_PROTO_SRC_FMT)]
+    
+    out_files = [ctx.actions.declare_file(out) for out in outs]
+    dir_out = _join_directories([
+        str(ctx.genfiles_dir.path), label_package, _GENERATED_PROTOS_DIR
+    ])
+
+    arguments = []
+    if ctx.executable.plugin:
+        arguments += get_plugin_args(
+            ctx.executable.plugin,
+            [],
+            dir_out,
+            False,
+        )
+        tools = [ctx.executable.plugin]
+    arguments += ["--objc_out=" + dir_out]
+
+    arguments += ["--proto_path=."]
+    arguments += get_include_protoc_args(protos)
+    # Include the output directory so that protoc puts the generated code in the
+    # right directory.
+    arguments += ["--proto_path={}".format(dir_out)]
+    arguments += ["--proto_path={}".format(_get_directory_from_proto(proto)) for proto in protos]
+    arguments += [_get_srcs_file_path(proto) for proto in protos]
+
+    # create a list of well known proto files if the argument is non-None
+    well_known_proto_files = []
+    if ctx.attr.use_well_known_protos:
+        f = ctx.attr.well_known_protos.files.to_list()[0].dirname
+        arguments += ["-I{0}".format(f + "/../..")]
```

Comment:
Can you explain a bit why you need to add `/../..`? (maybe with comments here)

Reply:
I guess it's for `google/protobuf`?

Reply:
Exactly